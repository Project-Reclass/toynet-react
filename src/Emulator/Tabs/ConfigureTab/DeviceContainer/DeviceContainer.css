.device-container {
  /* width: 21vw; */
  background-color: #212529;
  border-radius: 10px;
  margin: 0 0.5vw;
  overflow: auto;
  display: flex;
  z-index: 1;
  max-width: calc((100vw - 6.5vw - 15vw) / 3)
}


.device-container::-webkit-scrollbar {
  width: 5px;
  color: #595D62;
}

.device-container::-webkit-scrollbar-track {
  border-radius: 10px;
}

.device-container::-webkit-scrollbar-thumb {
  border-radius: 10px;
}

.device-container::-webkit-scrollbar-corner {
  border-radius: 10px;
  color: #595D62;
} 

.device-container-name {
  background-color: #454950;
  height: 4vh;
  line-height: 4vh;
  font-size: large;
  text-align: center;
  color: white;
  z-index: 9900;
  position: relative;
}

.device-container-content {
  color: white;
  height: 18vh;
  width: 100%;
  max-width: inherit;
  padding: 0.61375rem 0.4rem;
  padding-left: calc(0.61375rem + 5px);
  display: flex;
  flex-direction: column;
  overflow-y: scroll;
}

.plus-circle {
  height: auto;
  width: 18px;
  float: left;
  margin-top: 0.7vh;
  margin-left: 0.7vh;
  cursor: pointer;
  user-select: none;
  transition: 0.1s background-color ease-in-out;
  border-radius: 5px;
}

.plus-circle:hover {
  background-color: #3c841b;
}