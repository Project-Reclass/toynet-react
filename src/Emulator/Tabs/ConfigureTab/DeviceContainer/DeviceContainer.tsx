import React from 'react';
import styled from '@emotion/styled';

import { ReactComponent as PlusIcon } from 'src/assets/add.svg';

import { DeviceInterface } from 'src/common/types';
import { TopologyActions } from 'src/Emulator/useTopology';
import { useEmulator } from 'src/Emulator/EmulatorProvider';

import Device from './Device';

const DeviceContainerDiv = styled.div`
  background-color: #212529;
  border-radius: 10px;
  margin: 0 0.5vw;
  overflow: auto;
  display: flex;
  z-index: 1;
  max-width: calc((100vw - 6.5vw - 15vw) / 3);

  ::-webkit-scrollbar {
    width: 5px;
    color: #595D62;
  }

  ::-webkit-scrollbar-track {
    border-radius: 10px;
  }

  ::-webkit-scrollbar-thumb {
    border-radius: 10px;
  }

  ::-webkit-scrollbar-corner {
    border-radius: 10px;
    color: #595D62;
  }
`;

const DeviceContainerName = styled.div`
  height: 4vh;
  line-height: 4vh;
  font-size: large;
  text-align: center;
  color: white;
  z-index: 9900;
  position: relative;
`;

const DeviceContainerContent = styled.div`
  color: white;
  height: 28vh;
  width: 100%;
  max-width: inherit;
  padding: 0.61375rem 0.4rem;
  padding-left: calc(0.61375rem + 5px);
  display: flex;
  flex-direction: column;
  overflow-y: scroll;
`;

const HoverablePlusIcon = styled(PlusIcon)`
  height: auto;
  width: 18px;
  float: left;
  margin-top: 0.7vh;
  margin-left: 0.7vh;
  cursor: pointer;
  user-select: none;
  transition: 0.1s background-color ease-in-out;
  border-radius: 5px;

  :hover {
    background-color: #3c841b;
  }
`;

interface Props {
  deviceName: string;
  devices: DeviceInterface[];
  addDevice: (prefix: string) => any;
}

const DeviceContainer = React.forwardRef<HTMLDivElement, Props>(
  ({ deviceName, devices, addDevice }, ref) => {
    const { dispatch } = useEmulator();

    return (
      <div>
        <DeviceContainerName>
          <HoverablePlusIcon
            data-testid={'plus-icon'}
            onClick={() => addDevice(deviceName[0].toLowerCase())}
          />
          {deviceName}
        </DeviceContainerName>
        <DeviceContainerDiv>
          <DeviceContainerContent>
            {devices.map(device => (
              <Device
                key={`${device.name}`}
                deviceName={deviceName}
                deviceData={device}
                onDrop={(from, to) => dispatch({ type: TopologyActions.ADD_CONNECTION, payload: { to, from } })}
                onRemove={(from, to) => dispatch({ type: TopologyActions.DELETE_CONNECTION, payload: { to, from } })}
              />
            ))}
            <div ref={ref} />
          </DeviceContainerContent>
        </DeviceContainerDiv>
      </div>
    );
  },
);

export default DeviceContainer;
