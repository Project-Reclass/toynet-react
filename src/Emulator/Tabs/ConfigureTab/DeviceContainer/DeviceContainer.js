import React from "react";
import "./DeviceContainer.css";
import Device from "./Device/Device";

const DeviceContainer = ({ deviceName, devices }) => {
  return (
    <div className="device-container">
      <div className="device-container-name">{deviceName}</div>
      <div className="device-container-content">
        {devices.map((device) => (
          <Device deviceName={deviceName} deviceData={device} />
        ))}
      </div>
    </div>
  );
};

export default DeviceContainer;
