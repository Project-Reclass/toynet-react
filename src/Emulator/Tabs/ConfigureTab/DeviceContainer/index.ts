import DeviceContainer from './DeviceContainer';
export default DeviceContainer;