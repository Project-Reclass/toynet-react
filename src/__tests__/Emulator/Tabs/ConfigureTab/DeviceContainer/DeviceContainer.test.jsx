import React from 'react';
import renderer from 'react-test-renderer';

import DeviceContainer from 'src/Emulator/Tabs/ConfigureTab/DeviceContainer';
import { render, fireEvent } from '@testing-library/react';

const addDeviceMock = jest.fn();

const defaultProps = {
  deviceName: 'Router',
  devices: [
    {
      name: 'R1',
      connections: ['h1', 'h2', 's1']
    },
    {
      name: 'R2',
      connections: ['h3', 's1']
    }
  ],

  addDevice: addDeviceMock,
};

describe('The DeviceContainer', () => {
  afterEach(() => {
    addDeviceMock.mockClear();
  });

  it('should match previous snapshots', () => {
    const tree = renderer.create(<DeviceContainer {...defaultProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should add device when plus icon is clicked', () => {
    const { getByAltText } = render(<DeviceContainer {...defaultProps} />);
    const plusIcon = getByAltText(/plus icon/i);

    fireEvent.click(plusIcon);

    expect(addDeviceMock).toHaveBeenCalled();
    expect(addDeviceMock).toHaveBeenCalledWith('r');
  });
});