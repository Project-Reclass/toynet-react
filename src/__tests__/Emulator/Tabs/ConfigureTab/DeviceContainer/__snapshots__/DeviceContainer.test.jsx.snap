// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The DeviceContainer should match previous snapshots 1`] = `
<div>
  <div
    className="css-vl1v3j"
  >
    <svg
      className="css-8s9z6t"
      data-testid="plus-icon"
      onClick={[Function]}
    >
      add.svg
    </svg>
    Router
  </div>
  <div
    className="css-zop8l2"
  >
    <div
      className="css-ixj81f"
    >
      <div
        className="css-sdnyl6"
      >
        <div
          className="css-k008qs"
        >
          <div
            className="css-1fpqswo"
            size="large"
            style={
              Object {
                "margin": "auto",
              }
            }
            type="r"
          >
            R1
          </div>
          <hr
            aria-orientation="vertical"
            className="css-1re51h6"
          />
        </div>
        <div
          className="css-44y88i"
        >
          <div>
            Connections:
          </div>
          <div
            className="css-7sa7d"
          >
            <div
              className="css-r63nom"
              id="h1"
              style={
                Object {
                  "transform": "rotateX(-180deg)",
                }
              }
              type="h"
            >
              h1
            </div>
            <div
              className="css-r63nom"
              id="h2"
              style={
                Object {
                  "transform": "rotateX(-180deg)",
                }
              }
              type="h"
            >
              h2
            </div>
            <div
              className="css-qckvoc"
              id="s1"
              style={
                Object {
                  "transform": "rotateX(-180deg)",
                }
              }
              type="s"
            >
              s1
            </div>
          </div>
        </div>
        <div
          className="css-1y4kgfw"
          data-testid="trash-icon"
        >
          <hr
            aria-orientation="vertical"
            className="css-1re51h6"
          />
          <div
            className="css-1tp6xpy"
          >
            <svg
              className="css-t0psiy"
              focusable={true}
              role="presentation"
              viewBox="0 0 24 24"
            >
              <g
                fill="currentColor"
              >
                <path
                  d="M19.452,7.5H4.547a.5.5,0,0,0-.5.545L5.334,22.181A2,2,0,0,0,7.326,24h9.347a2,2,0,0,0,1.992-1.819L19.95,8.045a.5.5,0,0,0-.129-.382A.5.5,0,0,0,19.452,7.5Zm-9.2,13a.75.75,0,0,1-1.5,0v-9a.75.75,0,0,1,1.5,0Zm5,0a.75.75,0,0,1-1.5,0v-9a.75.75,0,0,1,1.5,0Z"
                />
                <path
                  d="M22,4H17.25A.25.25,0,0,1,17,3.75V2.5A2.5,2.5,0,0,0,14.5,0h-5A2.5,2.5,0,0,0,7,2.5V3.75A.25.25,0,0,1,6.75,4H2A1,1,0,0,0,2,6H22a1,1,0,0,0,0-2ZM9,3.75V2.5A.5.5,0,0,1,9.5,2h5a.5.5,0,0,1,.5.5V3.75a.25.25,0,0,1-.25.25H9.25A.25.25,0,0,1,9,3.75Z"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div
        className="css-sdnyl6"
      >
        <div
          className="css-k008qs"
        >
          <div
            className="css-1fpqswo"
            size="large"
            style={
              Object {
                "margin": "auto",
              }
            }
            type="r"
          >
            R2
          </div>
          <hr
            aria-orientation="vertical"
            className="css-1re51h6"
          />
        </div>
        <div
          className="css-44y88i"
        >
          <div>
            Connections:
          </div>
          <div
            className="css-7sa7d"
          >
            <div
              className="css-r63nom"
              id="h3"
              style={
                Object {
                  "transform": "rotateX(-180deg)",
                }
              }
              type="h"
            >
              h3
            </div>
            <div
              className="css-qckvoc"
              id="s1"
              style={
                Object {
                  "transform": "rotateX(-180deg)",
                }
              }
              type="s"
            >
              s1
            </div>
          </div>
        </div>
        <div
          className="css-1y4kgfw"
          data-testid="trash-icon"
        >
          <hr
            aria-orientation="vertical"
            className="css-1re51h6"
          />
          <div
            className="css-1tp6xpy"
          >
            <svg
              className="css-t0psiy"
              focusable={true}
              role="presentation"
              viewBox="0 0 24 24"
            >
              <g
                fill="currentColor"
              >
                <path
                  d="M19.452,7.5H4.547a.5.5,0,0,0-.5.545L5.334,22.181A2,2,0,0,0,7.326,24h9.347a2,2,0,0,0,1.992-1.819L19.95,8.045a.5.5,0,0,0-.129-.382A.5.5,0,0,0,19.452,7.5Zm-9.2,13a.75.75,0,0,1-1.5,0v-9a.75.75,0,0,1,1.5,0Zm5,0a.75.75,0,0,1-1.5,0v-9a.75.75,0,0,1,1.5,0Z"
                />
                <path
                  d="M22,4H17.25A.25.25,0,0,1,17,3.75V2.5A2.5,2.5,0,0,0,14.5,0h-5A2.5,2.5,0,0,0,7,2.5V3.75A.25.25,0,0,1,6.75,4H2A1,1,0,0,0,2,6H22a1,1,0,0,0,0-2ZM9,3.75V2.5A.5.5,0,0,1,9.5,2h5a.5.5,0,0,1,.5.5V3.75a.25.25,0,0,1-.25.25H9.25A.25.25,0,0,1,9,3.75Z"
                />
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div />
    </div>
  </div>
</div>
`;
